test/unit/node/app.test.ts

```diff
@@ -1,6 +1,262 @@
+import { logger } from "@coder/logger"
+import { promises, rmdirSync } from "fs"
 import * as http from "http"
-import { ensureAddress } from "../../../src/node/app"
-import { getAvailablePort } from "../../utils/helpers"
+import * as https from "https"
+import * as path from "path"
+import { createApp, ensureAddress, handleArgsSocketCatchError, handleServerError } from "../../../src/node/app"
+import { OptionalString, setDefaults } from "../../../src/node/cli"
+import { getAvailablePort, tmpdir } from "../../utils/helpers"
+
+// wip: todo@jsjoeio fix
+const FAKE_CRT = `-----BEGIN CERTIFICATE-----
+MIIFjjCCA3YCCQCVV/TwiFnSozANBgkqhkiG9w0BAQsFADCBiDELMAkGA1UEBhMC
+VVMxDjAMBgNVBAgMBVRleGFzMQ8wDQYDVQQHDAZBdXN0aW4xDjAMBgNVBAoMBUNv
+ZGVyMRQwEgYDVQQLDAtFbmdpbmVlcmluZzESMBAGA1UEAwwJY29kZXIuY29tMR4w
+HAYJKoZIhvcNAQkBFg9oZWxsb0Bjb2Rlci5jb20wHhcNMjEwOTE1MTgyNDIwWhcN
+MzEwOTEzMTgyNDIwWjCBiDELMAkGA1UEBhMCVVMxDjAMBgNVBAgMBVRleGFzMQ8w
+DQYDVQQHDAZBdXN0aW4xDjAMBgNVBAoMBUNvZGVyMRQwEgYDVQQLDAtFbmdpbmVl
+cmluZzESMBAGA1UEAwwJY29kZXIuY29tMR4wHAYJKoZIhvcNAQkBFg9oZWxsb0Bj
+b2Rlci5jb20wggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQCoczbPZMn4
+ILxg/Vw6512NgORHpMmfxSxqRORw3VM9bnxTiq1xPqP8ol/3Bvk7PGmvlNN3alcE
+vEapyV6ceF9FxCJARBXpIqNS/tLNtAjIIiCOrBjVXTrDI8jsYIYgojDue+i/OAxA
+0CPGXXZKPH6yvpjbczHZDrG5GYchWWMzA1gzq4iiQJH1MZviQS3tSZSqw0Vn1jYK
+V4BAOkSkxS4NrevjK7bIW9Dfrly9bfJfd+YZNr0ARp46G6Ke33xxF2KYpngscEFZ
+f5RYJLevFFDs1+UGn4/YTDcGv+6yC4ftz0OuM7CZqcc8UtE9g1yIm5kCllPVptW1
+6c8K8GC3l0kCDO3X5C0cWee86Hzlb+uIdoHTsjif6l9+v/FfmiRjeEDKcgh1fEgB
+z21GFsCq6XPPU975V38AftSxQ9RBs/Zfv7Ra+fFITISgdfQWH/pETrevsr1KUEf1
+xOkEZhEuQWo4XzKSNojFHRr7lvxrCUPGt9gIr5k3frxx5mVTqKJSiyjFTrFRCQ0n
+SlNlB5GzQEzPRMqGGxT+CjXz2Sgc0n0xbxm6tORYbKRygPTwSyaiZbRJQrYP/J0K
+XsZvKyQDU5HIzI3sUK3pnMbEhM08wptbRXjBGVtZ4etYfPhOUJogZ1VnXakBQHjg
+iGWQ2Jcd56FhdrPl9V/5AqRaly895bXM7QIDAQABMA0GCSqGSIb3DQEBCwUAA4IC
+AQCoWQyoeJLXXxy74uTYM9wng5O03p0IM7fnViPtYa/5oXjX/h2e6X3ijStJ6JmL
+mUL5Z7wY9JfHCbaKlC0ZrftwGrbldAPsKK1ZjOg9lgh1ZzkL++Akept992IEH9GU
+VB2l1N87xiQs7y40bwzlMbDmV/X9P2GhgyWpcUY8s6fkbczHU0tYJnOjj9H1HAlX
+yUsq587Z9j5TwF3f+UcprqEDaqwRhCmknXMJTbNeoL5T2d5WxXqZEiWJGD364nJw
+xt+VoriCwWZclltUcBqnDheyxmWWa6SGxz2aiS2cHC14V7sgaxpe05uSumGtzLeL
+H+x8Qx9a+eyNq/miV9czK++miaLyGQRkSDE35iFWUKpa1C1BcKl53uKfYCo6u83a
+nAWr/CATnyFgVZzMEJloSvYfnuA2+klJtZ9cRV0lp4CxefR+AKnkcNtmCr71LX12
+b7XTXHw68gEQp/iVZc0cP23gaDs26KIduLez2moF+iTjqn4J77epV6vGjuj27WL3
+hK8Z7KUFqUU6DSVJ2Z/qGBw/ZP78l//o2Jbu4KFfg5UviUzrLXWfrMJw9uqgfD4c
+OiahmBsqB/9JHGFGOhQy1/F8k/7Y3k1o1hdYwNCmB7iwPJ9eklhYHCrQc/yIarZz
+Z4rwv90aq+cFtoVvrTPA0rjSu4jwo2tI/nYfdZOfcCXhcw==
+-----END CERTIFICATE-----
+`
+
+const FAKE_KEY = `-----BEGIN PRIVATE KEY-----
```

Comment:
FYI these sorts of keys might cause problems with security scan tools. Might want to generate a small (e.g. 1024-bit) key every time the tests start

Reply:
Ahh! Nice catch. 

> Might want to generate a small (e.g. 1024-bit) key every time the tests start

I wonder how easy that is 🤔 I think when I made these locally, it ran me through an interactive prompt. I'll look into this.

Reply:
Try this instead of using OpenSSL directly to generate keys: https://www.npmjs.com/package/pem -- with very weak key settings, generating keys is very fast. We do something similar with our monorepo tests, for the same reason (to avoid having hardcoded keys flagged as a potential security issue)

Reply:
I also came across this: https://nodejs.org/api/tls.html#tls_tls_createsecurepair_context_isserver_requestcert_rejectunauthorized_options but I don't know Node.js well enough to know if this is helpful. You can also just merge this as-is and see if scanners pick it up or not, I don't think it's a big deal since this is just test code. It might be that Node.js doesn't provide any straightforward way to create x509 certificates for TLS

Reply:
Well, I'll at least take a look at both of those and see. Dude, thank you so much!

Reply:
Oh, looks like Node.js 15 introduced a new API; see: https://nodejs.org/api/crypto.html#crypto_new_x509certificate_buffer and https://www.geeksforgeeks.org/node-js-new-x509certificate-function/

Reply:
Turns out we had `pem` already and even a `generateCertificate` function 😅 Ended up using that. (Would have never thought to look without your suggestion so thank you ♥️)

https://github.com/cdr/code-server/blob/main/src/node/util.ts#L103-L146